### PR TITLE
Ww3 feature/before after fix

### PR DESCRIPTION
**Pull request by pstaabp**
This is a fix for issue #470.  

To check, there is a new line in webwork.apache2-config.dist (and webwork.apache2.4-config.dist) that adds the line: `SetEnv MOD_PERL_API_VERSION 2`

You will need to add this line to your non .dist version and restart the server.  You shouldn't get the warning in issue #470 anymore. 

I figure this is safe because if anyone is running apache2.2 or 2.4 then they will be running mod_perl2.  This variable gets picked up in Dancer that wasn't being set another way.

### CODE REVIEW

File: lib/WeBWorK/Utils.pm [hunk 1]
```diff
@@ -39,7 +39,7 @@ use Errno;
 use File::Path qw(rmtree);
 use Storable;
 use Carp;
-use Dancer qw(:script);
+#use Dancer qw(:script);
```
Comment (goehle) [inline lib/WeBWorK/Utils.pm]: Commenting this out is causing app.pl to return the error
```
Can't locate mod_perl.pm in @INC (@INC contains: /opt/webwork/pg/lib /opt/webwork/webwork2/lib /opt/webwork/webwork2-vanilla/webwork3/lib /opt/webwork/webwork2/webwork3/lib /etc/perl /usr/local/lib/perl/5.14.2 /usr/local/share/perl/5.14.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.14 /usr/share/perl/5.14 /usr/local/lib/site_perl .) at /opt/webwork/webwork2/lib/WeBWorK/Cookie.pm line 35.
```

**Comment by pstaabp** [comment]
Did you add the line:
```
SetEnv MOD_PERL_API_VERSION 2
```

in your webwork.apache2-config file?

**Comment by goehle** [comment]
Yes, that is the line that causes the error when it is present.

**Comment by pstaabp** [comment]
Where the error is occurring is in WeBWorK::Cookie because it is used via the Authen.pm module, which Dancer uses to login.  However, Dancer doesn't use mod_perl and when it hits line 35 of WeBWorK cookie, then it defaults to Apache1 and tries to load mod_perl.pm (instead of mod_perl2.pm).  What I thought would work is the line:
```
SetEnv MOD_PERL_API_VERSION 2
```
would send the cookie to the correct branch.  Evidently, it is not grabbing that environmental variable correctly.  

As I mentioned before, I stumbled on adding Dancer to the lib/WeBWorK/Utils.pm to not get the error, but it shouldn't be there.  That Utils.pm module has nothing to do with Dancer. 

I also tried adding a line like:
```
$ENV{MOD_PERL_API_VERSION } = 2;
```
inside the <Perl> part of webwork.apache2.4-config, but this didn't work on my production server. 

Any thoughts?

**Comment by pstaabp** [comment]
So this is working on my machine anymore.  I don't know what's going on.  For some reason, it's not picking up the env variable `$ENV{MOD_PERL_API_VERSION}`. 

I'll keep working on this.

**Comment by pstaabp** [comment]
It appears that environmental variables don't pass well into dancer applications.  @goehle: How are you deploying dancer to test ww3?

**Comment by goehle** [comment]
I'm using the fcgi setup.  That is going to be *super* bad at passing environment variables.  You might be able to add them directly to the fcgi script.

**Comment by goehle** [comment]
I still have problems with your latest changes.  Should we set the environment variable in app.pl?  It seem like that would be a catch all way to make sure it gets set by something running dancer.

**Comment by goehle** [comment]
OK.  I think I have something.  First, I set the environment variable in app.pl by adding 
```
BEGIN {$ENV{MOD_PERL_API_VERSION}=2};
```
*before* any of the `use WeBWorK::...` lines.  This sets the environment variable for all of the perl stuff correctly.  Unfortunately for things that call `use mod_perl` the mod_perl stuff is looking for a more deeply defined environment variable to decide if its supposed to be version 2.0 or not and its not picking up this environment variable.  So I removed the `use mod_perl` line from `Cookie.pm` and changed the begin block to 
```
BEGIN {
	if (MP2) {
	    require mod_perl2;
	    require Apache2::Cookie;
	    require APR::Request::Error;
	    Apache2::Cookie->import;
	    push @WeBWorK::Cookie::ISA, "Apache2::Cookie";
	} else {
	    require mod_perl;
	    require Apache::Cookie;
	    Apache::Cookie->import;
	    push @WeBWorK::Cookie::ISA, "Apache::Cookie";
	}
}
```
This explicitly uses the webwork MP2 variable (which is set correctly) to determine if mod_perl or mod_perl2 should be imported.  With this I was able to get ww3 running without including Dancer into Utils.

**Comment by pstaabp** [comment]
I deleted the remote branch to throw away a lot of the attempts. I thought I could reopen this.

**Comment by pstaabp** [comment]
I just opened a new version of this.  Thanks for the fix Geoff.  I almost had the exact code except didn't put the BEGIN block above the use WeBWorK lines and that was it.

**Pull request closed without merging**